extra_data/caldata.py

```diff
@@ -0,0 +1,1016 @@
+            snapshot_at (datetime, date, str or None): Time of database
+                state to look at.
+            **condition_params: Additional operating condition
+                parameters defined only for this query, ignored if
+                condition is passed.
+
+        Returns:
+            (CCVMetadata) CCV metadata result.
+        """
+
+        try:
+            dark_condition = condition[0]
+            illum_condition = condition[1]
+        except (TypeError, IndexError, AssertionError):
+            if condition is not None:
+                # If the condition was actually specified, inform the
+                # caller that this method's interface is different.
+                raise ValueError('condition must be 2-len sequence of dark '
+                                 'and illuminated condition')
+
+            dark_condition = self.dark_condition(**condition_params)
+            illum_condition = self.illuminated_condition(**condition_params)
+
+        if calibrations is None:
+            calibrations = (
+                self.dark_calibrations | self.illuminated_calibrations)
+
+        metadata = CCVMetadata()
+
+        dark_calibrations = self.dark_calibrations & set(calibrations)
+        if dark_calibrations:
+            self._retrieve_constant_version(
+                metadata, dark_calibrations, dark_condition,
+                event_at, snapshot_at)
+
+        illum_calibrations = self.illuminated_calibrations & set(calibrations)
+        if illum_calibrations:
+            self._retrieve_constant_version(
+                metadata, illum_calibrations, illum_condition,
+                event_at, snapshot_at)
+
+        return metadata
+
+
+class AGIPD_CalibrationData(DualConditionCalibrationData):
+    """Calibration data for the AGIPD detector.
+
+    Dark operating condition:
+        Required parameters:
+            * Sensor Bias Voltage: 300.0
+            * Pixels X: 512
+            * Pixels Y: 128
+
+        Optional parameters with default values:
+            * Memory cells: 202
+            * Acquisition rate: 4.5
+
+        Optional parameters:
+            * Gain setting
+            * Gain mode
+            * Integration time
+
+    Illuminated operating condition:
+        Required parameter:
+            * Source energy: 9.2
+
+        and all of the above for the dark condition.
+    """
+
+    dark_calibrations = {'Offset', 'Noise', 'ThresholdsDark', 'BadPixelsDark',
+                         'BadPixelsPC', 'SlopesPC'}
+    illuminated_calibrations = {'BadPixelsFF', 'SlopesFF'}
+
+    def __init__(self, detector_name, modules=None, client=None,
+                 **condition_params):
+        super().__init__(detector_name, modules=modules, client=client,
+                         **condition_params)
+
+    @classmethod
+    def from_data(cls, data, detector=None, modules=None, client=None,
+                  ctrl_device_id=None, **condition_params):
```

Comment:
I think all these methods that take `**condition_params` will accept and ignore any keyword argument, so if I make a typo like `intergation_time`, the value I pass will be silently ignored? I think we should error on unrecognised keyword arguments, to catch this kind of mistake.